modules/okhttp-client/src/main/java/com/spotify/apollo/http/client/LoggingHttpClient.java

```diff
@@ -0,0 +1,107 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.apollo.http.client;
+
+import com.spotify.apollo.Request;
+import com.spotify.apollo.Response;
+import com.spotify.apollo.environment.IncomingRequestAwareClient;
+import com.spotify.metrics.core.MetricId;
+import com.spotify.metrics.core.SemanticMetricRegistry;
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+import javax.annotation.Nullable;
+import okio.ByteString;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LoggingHttpClient implements IncomingRequestAwareClient {
+
+  private static final Logger logger = LoggerFactory.getLogger(LoggingHttpClient.class);
+
+  private final IncomingRequestAwareClient client;
+  private final SemanticMetricRegistry metricRegistry;
+  private final MetricId baseId = MetricId.build();
+
+  public LoggingHttpClient(
+      IncomingRequestAwareClient client, SemanticMetricRegistry metricRegistry) {
+    this.client = client;
+    this.metricRegistry = metricRegistry;
+  }
+
+  @Override
+  public CompletionStage<Response<ByteString>> send(Request request, Optional<Request> incoming) {
+
+    if (!request.uri().startsWith("http")) {
+      return client.send(request, incoming);
+    }
+
+    final long start = System.currentTimeMillis();
+
+    return client
+        .send(request, incoming)
+        .whenComplete(
+            (response, throwable) -> {
+              final long elapsedMillis = System.currentTimeMillis() - start;
+              if (throwable != null) {
+                logOutgoingRequest(request, elapsedMillis, null);
+                logger.error("Exception caught: ", throwable);
+              } else {
+                logOutgoingRequest(request, elapsedMillis, response);
+              }
+            });
+  }
+
+  private void logOutgoingRequest(
+      final Request request, final long elapsedMillis, @Nullable final Response<?> response) {
+    final String statusCode =
+        response != null ? Integer.toString(response.status().code()) : "EXCEPTION";
+
+    final String targetHost = getHost(request);
+
+    metricRegistry
+        .meter(
+            baseId
+                .tagged("what", "message-rate")
+                .tagged("status-code", statusCode)
+                .tagged("target-host", targetHost)
+                .tagged("protocol", "http/https")
+                .tagged("unit", "request"))
+        .mark();
+
+    metricRegistry
+        .histogram(
+            baseId
+                .tagged("what", "request-latency")
+                .tagged("unit", "ms")
+                .tagged("protocol", "http/https")
+                .tagged("target-host", targetHost))
+        .update(elapsedMillis);
+  }
+
+  private String getHost(Request request) {
+    try {
+      return URI.create(request.uri()).getHost();
```

Comment:
I think this could just be `request.httpUrl().host()`, without having to reparse a URL or handle an exception

https://github.com/square/okhttp/blob/98a2ce65a424a700cf08d76df9d8e7a4afe42eff/okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java#L390

Reply:
Hmm, but `Request` interface has no direct `httpUrl()` so it will still need to be built

Reply:
Sorry, I thought this was the okhttp `Request` class 🤦‍♂️  please disregard